steps/s3/session.go

```diff
@@ -0,0 +1,207 @@
+	logger.LogMessage(fmt.Sprintf("validating the existence of bucket: %s", bucket))
+	cparams := &s3.CreateBucketInput{
+		Bucket: aws.String(bucket),
+	}
+
+	s3Client := s3.New(s.Client)
+	_, err := s3Client.CreateBucket(cparams)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == s3.ErrCodeBucketAlreadyExists || aerr.Code() == s3.ErrCodeBucketAlreadyOwnedByYou {
+				logger.LogMessage(fmt.Sprintf("validated the existence of bucket: %s", bucket))
+				return nil
+			}
+		} else {
+			return fmt.Errorf("error validating the existence of bucket: %s", bucket)
+		}
+	}
+	return fmt.Errorf("bucket: '%s' does not exists", bucket)
+}
+
+// ValidateS3File checks the existence of a file in S3.
+func (s *Session) ValidateS3FileExists(ctx context.Context, bucket, key string) error {
+	logger := GetLogger()
+	logger.LogOperation("validate", bucket, key)
+	s3svc := s3.New(s.Client)
+	exists, err := s.s3KeyExists(s3svc, bucket, key)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return fmt.Errorf("failed validating s3 file exits: no file exists in bucket '%s' with name '%s' ", bucket, key)
+	}
+	return nil
+}
+
+// ValidateS3FileWithContent checks the existence of a file in S3 with the content specified.
+func (s *Session) ValidateS3FileExistsWithContent(ctx context.Context, bucket, key, message string) error {
+	expected := strings.TrimSpace(message)
+	logger := GetLogger()
+	logger.LogOperation("validate", bucket, key)
+	downloader := s3manager.NewDownloader(s.Client)
+	buf := aws.NewWriteAtBuffer([]byte{})
+	_, err := downloader.Download(buf, &s3.GetObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		return fmt.Errorf("unable to upload %q to %q, %v", key, bucket, err)
+	}
+	actual := strings.TrimSpace(string(buf.Bytes()))
+	if expected != actual {
+		return fmt.Errorf("failed validating s3 bucket '%s' file '%s' content: expected:\n%s\n\nactual:\n%s", bucket, key, expected, actual)
+	}
+	return nil
+}
+
+// DeleteS3File deletes the file in S3.
+func (s *Session) DeleteS3File(ctx context.Context, bucket, key string) error {
+	logger := GetLogger()
+	logger.LogOperation("delete", bucket, key)
+	s3svc := s3.New(s.Client)
+	input := &s3.DeleteObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	}
+	_, err := s3svc.DeleteObject(input)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// s3KeyExists checks the existence of a key in a S3 bucket.
+func (s *Session) s3KeyExists(s3svc *s3.S3, bucket string, key string) (bool, error) {
+	_, err := s3svc.HeadObject(&s3.HeadObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case "NotFound": // s3.ErrCodeNoSuchKey does not work, aws is missing this error code so we hardwire a string
+				return false, nil
+			default:
+				return false, err
+			}
+		}
+		return false, err
+	}
+	return true, nil
```

Comment:
Without nesting error:
```
                 if err == nil {
                          return true, nil
                }
		if aerr, ok := err.(awserr.Error); !ok {
		         return false, err
		}
		return false, nil
```

Reply:
But that does not check if the error given by S3 is a not found error

Reply:
so that you could declare `var aerr error` and check it in the end:

                if err == nil {
                          return true, nil
                }
                var aerr error
                var ok bool
		if aerr, ok = err.(awserr.Error); !ok {
		         return false, err
		}
                if aerr.Code() == "NotFound" {
                         // Log it.
               }
	       return false, nil

Reply:
Done